orm/src/test/java/org/hibernate/search/test/batchindexing/IdentifierFilterTest.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.test.batchindexing;
+
+import java.util.Map;
+
+import org.hibernate.cfg.AvailableSettings;
+import org.hibernate.criterion.Criterion;
+import org.hibernate.criterion.DetachedCriteria;
+import org.hibernate.criterion.Restrictions;
+import org.hibernate.search.FullTextSession;
+import org.hibernate.search.MassIndexer;
+import org.hibernate.search.Search;
+import org.hibernate.search.batchindexing.spi.IdentifierCriteriaProvider;
+import org.hibernate.search.test.SearchTestBase;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test that illustrates how to specify {@link DetachedCriteria} filters for
+ * index types by calling {@link MassIndexer#registerCriteriaProvider(Class, IdentifierCriteriaProvider)}.
+ *
+ * @author Chris Cranford
+ */
+public class IdentifierFilterTest extends SearchTestBase {
+
+	@Test
+	public void testMassIndexerFilterSingleType() throws InterruptedException {
+		final FullTextSession fullTextSession = openAndPopulate();
+
+		getMassIndexer( fullTextSession, Clock.class )
+				.registerCriteriaProvider( Clock.class, new IdentifierCriteriaProvider() {
+					@Override
+					public Criterion getIdentifierCriteria(Class<?> indexedType) {
+						return Restrictions.idEq( 2 );
+					}
+				} )
+				.startAndWait();
+
+		fullTextSession.close();
+
+		assertEquals( 1, getNumberOfDocumentsInIndex( Clock.class ) );
+	}
+
+	@Test
+	public void testMassIndexerFilterMultipleTypes() throws InterruptedException {
+		final FullTextSession fullTextSession = openAndPopulate();
+
+		getMassIndexer( fullTextSession, Clock.class, IssueEntity.class )
```

Comment:
So keep the original approach for backward compatibility, but provide a type-builder fashion as you described above, allowing for all 3 variants, like below?

    // original way - backward compatibility.
    fullTextSession.createIndexer( Entity1.class, Entity2.class ).startAndWait();
    
    // new way and old mixed
    fullTextSession.createIndexer( Entity2.class )
      .type( Entity1.class ) // delayed registration possible
      .startAndWait();

    // new way with identifier criteria provider
    fullTextSession.createIndexer( Entity2.class )
      .type( Entity1.class ) // delayed registration with identifier limit binder
        .limitedBy( newIdentifierCriteriaProvider() { /* things */ } )
      .startAndWait();